typings/styled-components.d.ts

```diff
@@ -35,13 +33,14 @@ type Attrs<P, A extends Partial<P>, T> = {
 export interface StyledComponentClass<P, T, O = P> extends ComponentClass<ThemedOuterStyledProps<O, T>> {
   extend: ThemedStyledFunction<P, T, O>;
 
-  withComponent<K extends keyof JSX.IntrinsicElements>(tag: K): StyledComponentClass<JSX.IntrinsicElements[K], T, O>;
-  withComponent(element: ComponentClass<P>): StyledComponentClass<P, T, O>;
+  withComponent<K extends keyof JSX.IntrinsicElements>(tag: K): StyledComponentClass<JSX.IntrinsicElements[K], T, JSX.IntrinsicElements[K]>;
+  withComponent<U extends Partial<P> = {}>(element: React.ComponentType<U>): StyledComponentClass<U & P, T, U & O>;
```

Comment:
I guess `U` is no need to extend `P` here,
but I think we need to compromise a bit here between newly passed component props `U` and base component props `P` here. 
In my point of view, returned component at lease need to have all props `U` since it is the component we want to use ?